src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java

```diff
@@ -64,70 +102,65 @@ void setProjectId(final String projectId) {
     if (!Files.exists(file)) {
       throw new MojoExecutionException(
           file
               + " does not exist. Make sure to bind the scan goal to a phase that will execute after the artifact to scan has been built");
     }
-    final NewCodeArtifactRequest codeArtifactRequest = NewCodeArtifactRequest.of(projectId, file);
-    getLog().info("Uploading " + file.getFileName() + " to Contrast Scan");
 
-    // create new code artifact
-    final CodeArtifact artifact;
+    final ContrastScanSDK contrastScan = new DefaultContrastScanSDK(contrast, getURL());
+    final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+    final ArtifactScanner scanner =
+        new ArtifactScanner(
+            executor, contrastScan, getOrganizationId(), getProjectId(), POLL_SCAN_INTERVAL);
     try {
-      artifact = contrastScan.createCodeArtifact(getOrganizationId(), codeArtifactRequest);
-    } catch (final UnauthorizedException e) {
-      throw new MojoExecutionException("Failed to authenticate to Contrast", e);
-    } catch (final IOException e) {
-      throw new MojoExecutionException("Failed to upload artifact to Contrast Scan", e);
-    }
+      getLog().info("Uploading " + file.getFileName() + " to Contrast Scan");
+      final ScanOperation operation = scanner.scanArtifact(file, label);
+      getLog().info("Starting scan with label " + label);
 
-    // start scan for new code artifact
-    final StartScanRequest request =
-        StartScanRequest.builder()
-            .projectId(projectId)
-            .codeArtifactId(artifact.getId())
-            .label(label)
-            .build();
-    final Scan scan;
-    try {
-      scan = contrastScan.startScan(getOrganizationId(), request);
-    } catch (final UnauthorizedException e) {
-      throw new MojoExecutionException("Failed to authenticate to Contrast", e);
-    } catch (final IOException e) {
-      throw new MojoExecutionException(
-          "Failed to start scan for code artifact " + artifact.getId(), e);
-    }
+      // show link in build log
+      final URL clickableScanURL = createClickableScanURL(operation.id());
+      getLog().info("Scan results will be available at " + clickableScanURL.toExternalForm());
 
-    // show link in build log
-    final URL clickableScanURL;
-    try {
-      clickableScanURL = createClickableScanURL(scan.getId());
-    } catch (final MalformedURLException e) {
-      throw new MojoExecutionException(
-          "Error building clickable Scan URL. Please contact support@contrastsecurity.com for help",
-          e);
+      // if should not wait, then stop asking for scan results and return
+      if (!waitForResults) {
+        operation.disconnect();
+        return;
+      }
+
+      // else wait for results, output summary to console, output sarif to file system
+      getLog().info("Waiting for scan results");
+      waitForResults(operation);
+    } finally {
+      executor.shutdown();
```

Comment:
Should we do `shutdownNow` if `waitForResults == false`?

Reply:
I'm not sure it matters. Do you think we should do shutdownNow() for both?

When `waitForResults == true`, and we call `shutdown`, the call to `shutdown` immediately returns and we cede control back to Maven. In the background, the `ExecutorService` is still running, and the only possible task that it's still running is that first poll to Contrast Scan API for the status of the scan. That should complete quickly at which point no further polling will happen.

When `waitForResults == false`, there should be no tasks running anyway, so `shutdown` and `shutdownNow()` are basically the same thing, but with the former we aren't compelled to double check the returned list of runnable I guess.

Thoughts?

Reply:
Yeah I think `shutdown` is fine. I wasn’t thinking about the fact that this is a `finally`. I was thinking of it as a `catch` where there’s been some error and potentially we should shutdown right away, but even then `shutdown` seems okay. I think we can leave it.